From 40484cb0b5034bc3c1e23b2ae1f2d39eedce07e9 Mon Sep 17 00:00:00 2001
From: Viktor Ashirov <vashirov@redhat.com>
Date: Tue, 27 Jan 2026 14:26:29 +0100
Subject: [PATCH] Issue 7096 - (2nd) During replication online total init the
 function idl_id_is_in_idlist is not scaling with large database (#7205)

Bug Description:
The fix for #7096 optimized the BDB backend's `idl_new_range_fetch()`
function to use ID ranges instead of checking the full ID list during
online total initialization. However, the LMDB backend's
`idl_lmdb_range_fetch()` function and its callback
`idl_range_add_id_cb()` were not updated and still use the non-scaling
`idl_id_is_in_idlist()` function.

Fix Description:
Apply the same optimization to the LMDB backend.

Fixes: https://github.com/389ds/389-ds-base/issues/7096

Reviewed by: @tbordaz, @droideck (Thanks!)
---
 ldap/servers/slapd/back-ldbm/idl_new.c | 39 ++++++++++----------------
 1 file changed, 15 insertions(+), 24 deletions(-)

diff --git a/ldap/servers/slapd/back-ldbm/idl_new.c b/ldap/servers/slapd/back-ldbm/idl_new.c
index 2d978353f..613d53815 100644
--- a/ldap/servers/slapd/back-ldbm/idl_new.c
+++ b/ldap/servers/slapd/back-ldbm/idl_new.c
@@ -66,6 +66,7 @@ typedef struct {
     size_t leftoverlen;
     size_t leftovercnt;
     IDList *idl;
+    IdRange_t *idrange_list;
     int flag_err;
     ID lastid;
     ID suffix;
@@ -700,9 +701,9 @@ error:
                 }
             }
         }
-        slapi_ch_free((void **)&leftover);
-        idrange_free(&idrange_list);
     }
+    slapi_ch_free((void **)&leftover);
+    idrange_free(&idrange_list);
     slapi_log_err(SLAPI_LOG_FILTER, "idl_new_range_fetch",
                   "Found %d candidates; error code is: %d\n",
                   idl ? idl->b_nids : 0, *flag_err);
@@ -716,7 +717,6 @@ static int
 idl_range_add_id_cb(dbi_val_t *key, dbi_val_t *data, void *ctx)
 {
     idl_range_ctx_t *rctx = ctx;
-    int idl_rc = 0;
     ID id = 0;
 
     if (key->data == NULL) {
@@ -779,10 +779,12 @@ idl_range_add_id_cb(dbi_val_t *key, dbi_val_t *data, void *ctx)
              * found entry is the one from the suffix
              */
             rctx->suffix = keyval;
-            idl_rc = idl_append_extend(&rctx->idl, id);
-        } else if ((keyval == rctx->suffix) || idl_id_is_in_idlist(rctx->idl, keyval)) {
+            idl_append_extend(&rctx->idl, id);
+            idrange_add_id(&rctx->idrange_list, id);
+        } else if ((keyval == rctx->suffix) || idl_id_is_in_idlist_ranges(rctx->idl, rctx->idrange_list, keyval)) {
             /* the parent is the suffix or already in idl. */
-            idl_rc = idl_append_extend(&rctx->idl, id);
+            idl_append_extend(&rctx->idl, id);
+            idrange_add_id(&rctx->idrange_list, id);
         } else {
             /* Otherwise, keep the {keyval,id} in leftover array */
             if (!rctx->leftover) {
@@ -797,14 +799,7 @@ idl_range_add_id_cb(dbi_val_t *key, dbi_val_t *data, void *ctx)
             rctx->leftovercnt++;
         }
     } else {
-        idl_rc = idl_append_extend(&rctx->idl, id);
-    }
-    if (idl_rc) {
-        slapi_log_err(SLAPI_LOG_ERR, "idl_lmdb_range_fetch",
-                      "Unable to extend id list (err=%d)\n", idl_rc);
-        idl_free(&rctx->idl);
-        rctx->flag_err = LDAP_UNWILLING_TO_PERFORM;
-        return DBI_RC_NOTFOUND;
+        idl_append_extend(&rctx->idl, id);
     }
 #if defined(DB_ALLIDS_ON_READ)
     /* enforce the allids read limit */
@@ -841,7 +836,6 @@ idl_lmdb_range_fetch(
 {
     int ret = 0;
     int ret2 = 0;
-    int idl_rc = 0;
     dbi_cursor_t cursor = {0};
     back_txn s_txn;
     struct ldbminfo *li = (struct ldbminfo *)be->be_database->plg_private;
@@ -891,6 +885,7 @@ idl_lmdb_range_fetch(
     idl_range_ctx.lastid = 0;
     idl_range_ctx.count = 0;
     idl_range_ctx.index_id = index_id;
+    idl_range_ctx.idrange_list = NULL;
     if (operator & SLAPI_OP_RANGE_NO_IDL_SORT) {
             struct _back_info_index_key bck_info;
             /* We are doing a bulk import
@@ -966,22 +961,18 @@ error:
         while(remaining > 0) {
             for (size_t i = 0; i < idl_range_ctx.leftovercnt; i++) {
                 if (idl_range_ctx.leftover[i].key > 0 &&
-                    idl_id_is_in_idlist(idl_range_ctx.idl, idl_range_ctx.leftover[i].key) != 0) {
+                    idl_id_is_in_idlist_ranges(idl_range_ctx.idl, idl_range_ctx.idrange_list, idl_range_ctx.leftover[i].key) != 0) {
                     /* if the leftover key has its parent in the idl */
-                    idl_rc = idl_append_extend(&idl_range_ctx.idl, idl_range_ctx.leftover[i].id);
-                    if (idl_rc) {
-                        slapi_log_err(SLAPI_LOG_ERR, "idl_lmdb_range_fetch",
-                                      "Unable to extend id list (err=%d)\n", idl_rc);
-                        idl_free(&idl_range_ctx.idl);
-                        break;
-                    }
+                    idl_append_extend(&idl_range_ctx.idl, idl_range_ctx.leftover[i].id);
+                    idrange_add_id(&idl_range_ctx.idrange_list, idl_range_ctx.leftover[i].id);
                     idl_range_ctx.leftover[i].key = 0;
                     remaining--;
                 }
             }
         }
-        slapi_ch_free((void **)&idl_range_ctx.leftover);
     }
+    slapi_ch_free((void **)&idl_range_ctx.leftover);
+    idrange_free(&idl_range_ctx.idrange_list);
     *flag_err = idl_range_ctx.flag_err;
     slapi_log_err(SLAPI_LOG_FILTER, "idl_lmdb_range_fetch",
                   "Found %d candidates; error code is: %d\n",
-- 
2.52.0

